### DISCUSSION TITLE
Secrets support in GitHub Codespaces

### DISCUSSION THREAD

**Discussion by johnboyes** [community/community#24964] [Codespaces]
<p>Hi,</p>
<p>would be great if GitHub Codespaces supported secrets, similar to how the now-deprecated <a href="https://github.com/MicrosoftDocs/vscodespaces/issues/50#issuecomment-652196750" rel="nofollow noopener">Visual Studio Codespaces does</a>.</p>
<p>I wonder if the <a href="https://docs.github.com/en/actions/configuring-and-managing-workflows/creating-and-storing-encrypted-secrets" rel="nofollow noopener">secrets functionality already in GitHub</a> (used by GitHub Actions) could also be integrated with GitHub Codespaces?</p>

**Comment by lostintangent** [comment]
<p>Hey! Secrets support is definitely on our roadmap :+1: As you mentioned, the plan is to leverage the existing support for org and repo-level secrets that Actions already uses. That said, since Codespaces is more of an individually-managed service (as opposed to Actions), we’ve also heard a lot of feedback for introducing user-level secrets.</p>
<p>Out of curiosity: how would you imagine your team making use of secrets with Codespaces? Would you have repo/org-wide secrets (e.g. connection strings for shared dev databases), or would you expect to want each developer to have individual, user-level secrets? Thanks!</p>

**Comment by johnboyes** [comment]
<p>Hi, thanks for the quick reply :slightly_smiling_face:  That’s great that it’s on the roadmap.</p>
<p>I think teams would find both repo/org-wide secrets <em>and</em> also individual, user-level secrets very useful.  I think it’s key that individual developers are able to configure secrets for their development environment on an individual basis.</p>
<p>As per <a href="https://12factor.net/config" rel="nofollow noopener">12 factor</a>, I’d see the individual level and the repo level secrets as being independent of each other, rather than trying to group them and to use overrides.  Would be very handy to be able to switch easily from having the Codespaces dev environment use individual secrets to using the repo level secrets, and vice versa.  I’d see the repo level secrets as typically being the default when creating a new Codespace, which would be useful for people new to a project, for instance.</p>
<p>Hope that’s useful, just throwing some ideas out there :slight_smile:</p>

**Comment by nickoneill** [comment]
<p>+1, being able to set environment vars from secrets would be ideal for being able to develop in codespaces.</p>

**Comment by miking-the-viking** [comment]
<p>+1 Strongly agreed around the secrets, especially personal.</p>
<p>From my understanding the ultimate philosophy of Codespaces is one-click, immediate Dev Environment. If we have any remaining manual setup after the automated setup steps then we’ve got a problem.</p>
<p>My current project necessitates using an OracleDB container, to pull that I need to login to Docker. That’s a private environment variable and as part of a developer’s contribution to the repo they will need to provide that.</p>
<p>When I create the codespace for this repo, it should spin up the entire stack, run migrations + seeders, install all needed dependencies, then by the time Codespaces is ready for me all I need to do is start coding. I’m fine to have more manual configuration upfront (re: secrets, build steps, etc.), I’d rather not have to repeat the process every time I made a Codespace (logging into DockerHub, running setup steps manually).</p>

**Comment by johnboyes** [comment]
<p>Hi <a class="mention" href="/lostintangent">@lostintangent</a>, just wondering if you have any updates on Secrets support - what’s the rough timeline for this on your roadmap?  Thanks v much, John</p>

**Comment by matthewisabel** [comment]
<p>We’re targeting having this available in a few weeks</p>

**Comment by ritz078** [comment]
<p>Is the secrets support there yet ?</p>

**Comment by rainloreley** [comment]
<p><a class="mention" href="/ritz078">@ritz078</a> they mentioned this question yesterday at GitHub Universe 2020, it should be available <em>soon</em></p>

**Comment by bailey** [comment, marked as answer]
<p><a class="mention" href="/johnboyes">@johnboyes</a> Hey John! We just announced secrets support if you want to check it out :eyes: Let us know what you think. <a href="https://github.community/t/new-features-in-github-codespaces/154887" class="inline-onebox">New Features in GitHub Codespaces</a></p>

**Comment by johnboyes** [comment]
<p>Thanks so much for thinking to let me (and this thread) know!  Will give it a whirl for sure and let you know how I get on :sunglasses:</p>

**Comment by johnboyes** [comment]
<p>Just tried using the new secrets functionality and am pleased to say it works a treat :blush:</p>

**Comment by ScottBrenner** [comment]
<p>Hey <a class="mention" href="/johnboyes">@johnboyes</a> - how did you actually <em>use</em> the secrets in your Codespace? I see <a href="https://docs.github.com/en/github/developing-online-with-codespaces/managing-encrypted-secrets-for-codespaces" class="inline-onebox" rel="noopener nofollow ugc">Managing encrypted secrets for Codespaces - GitHub Docs</a> for creating the secrets, but no documentation on how to then use/access it within a Codespace.</p>

**Comment by jkeech** [comment]
<p><a class="mention" href="/scottbrenner">@scottbrenner</a>, secrets end up as environment variables in the VS Code process and terminals inside the codespace. You should be able to run <code>echo $SECRET_NAME</code> in the terminal to see the value.</p>
<p>Secrets get updated on every codespace creation/resume, so if you update a secret value in the UI while the codespace is running, you’ll need to suspend and resume to pick up the updated value, either by letting it auto-suspend after 30 minutes of inactivity, or by using the VS Code extension to suspend the codespace manually and reconnect.</p>

**Comment by jtcressy** [comment]
<p>Are there any plans to expose these secrets as filesystem objects inside codespaces?<br>
The perfect use case for this would be GOOGLE_APPLICATION_CREDENTIALS for authenticating to google cloud. This environment variable is set to the <em>path</em> which contains the account’s json.</p>
<p>Ideally, this feature would work exactly like gitlab’s CI variables where you can designate a variable as a “file” and it will get placed in some location on the filesystem. Then, the path to that location is exposed in the environment variable.</p>

**Comment by byoung** [comment]
What about build secrets? Maybe I missed it in the docs, but how am I supposed to handle connecting to private package repos when building the container for a Codespace?
